services/dynamic-sidecar/tests/unit/test_api_containers.py

```diff
@@ -0,0 +1,283 @@
+    response = await test_client.post(f"/{api_vtag}/containers")
+    assert response.status_code == status.HTTP_500_INTERNAL_SERVER_ERROR, response.text
+    assert json.loads(response.text) == {"detail": "\nProvided yaml is not valid!"}
+
+
+async def test_compose_up_spec_invalid(test_client: TestClient) -> None:
+    invalid_compose_spec = Faker().text()
+    response = await test_client.post(
+        f"/{api_vtag}/containers", data=invalid_compose_spec
+    )
+    assert response.status_code == status.HTTP_500_INTERNAL_SERVER_ERROR, response.text
+    assert "Provided yaml is not valid!" in response.text
+    # 28+ characters means the compos spec is also present in the error message
+    assert len(response.text) > 28
+
+
+async def test_containers_down_after_starting(
+    test_client: TestClient, compose_spec: Dict[str, Any]
+) -> None:
+    # store spec first
+    response = await test_client.post(f"/{api_vtag}/containers", data=compose_spec)
+    assert response.status_code == status.HTTP_202_ACCEPTED, response.text
+    shared_store: SharedStore = test_client.application.state.shared_store
+    container_names = shared_store.container_names
+    assert json.loads(response.text) == container_names
+
+    response = await test_client.post(
+        f"/{api_vtag}/containers:down",
+        query_string=dict(command_timeout=DEFAULT_COMMAND_TIMEOUT),
+    )
+    assert response.status_code == status.HTTP_200_OK, response.text
+    assert response.text != ""
+
+
+async def test_containers_down_missing_spec(
+    test_client: TestClient, compose_spec: Dict[str, Any]
+) -> None:
+    response = await test_client.post(
+        f"/{api_vtag}/containers:down",
+        query_string=dict(command_timeout=DEFAULT_COMMAND_TIMEOUT),
+    )
+    assert response.status_code == status.HTTP_422_UNPROCESSABLE_ENTITY, response.text
+    assert json.loads(response.text) == {
+        "detail": "No spec for docker-compose down was found"
+    }
+
+
+def assert_keys_exist(result: Dict[str, Any]) -> bool:
+    for entry in result.values():
+        assert "Status" in entry
+        assert "Error" in entry
+    return True
+
+
+async def test_containers_get(
+    test_client: TestClient, started_containers: List[str]
+) -> None:
+    response = await test_client.get(f"/{api_vtag}/containers")
+    assert response.status_code == status.HTTP_200_OK, response.text
+
+    decoded_response = json.loads(response.text)
+    assert set(decoded_response) == set(started_containers)
+    for entry in decoded_response.values():
+        assert "Status" not in entry
+        assert "Error" not in entry
+
+
+async def test_containers_get_status(
+    test_client: TestClient, started_containers: List[str]
+) -> None:
+    response = await test_client.get(
+        f"/{api_vtag}/containers", query_string=dict(only_status=True)
+    )
+    assert response.status_code == status.HTTP_200_OK, response.text
+
+    decoded_response = json.loads(response.text)
```

Comment:
I think you can do response.json() everywhere instead of json.loads(text)

Reply:
yes, fixed that.